c/examples/cpp-experiment.cpp

```diff
@@ -0,0 +1,215 @@
+/* C++ wrapper experiement. */
+
+#include <iostream>
+#include <stdexcept>
+#include <sstream>
+#include <memory>
+#include <cassert>
+
+#include <tskit/tables.h>
+
+using namespace std;
+
+void
+check_error(int val)
+{
+    if (val < 0) {
+        std::ostringstream o;
+        o << tsk_strerror(val);
+        throw std::runtime_error(o.str());
+    }
+}
+
+namespace tskit {
+
+    class NodeTable
+    {
+
+        private:
+            tsk_node_table_t *table;
+            const bool allocated_locally;
+            using node_table_ptr = std::unique_ptr<tsk_node_table_t,void(*)(tsk_node_table_t*)>;
+
+            node_table_ptr make_empty_table()
+            {
+                node_table_ptr t(new tsk_node_table_t{},[](tsk_node_table_t* nt){delete nt;});
+                if (t.get() == NULL) {
```

Comment:
I did a bit of digging and I don't think we need to worry about ``new`` returning NULL: it raises an exception if allocation fails.

Reply:
Yeah.  This actually brings up a whole other issue :).  I thought that the C++ interface should be "new and delete only", but then I wondered about the idea of a TableCollection "releasing" its data to a bare pointer, in the manner of `unique_ptr::release`.  In that case, it would have to be a "malloc/free" combo.

There is a [non-throwing](http://www.cplusplus.com/reference/new/nothrow/) way to call new, which returns `nullptr`.  I wonder if I left that `NULL` check in there when switching away from malloc?  It should be `nullptr` and not `NULL`, which makes me think I left some cruft around.